Cpp_test/lib/ipc-shm-receive.cpp

```diff
@@ -0,0 +1,107 @@
+#include "ipc-shm-receive.h"
+
+IPCShmReceive::~IPCShmReceive()
+{
+	shm_unlink(this->info.method_name);
+}
+
+void IPCShmReceive::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+}
+
+void IPCShmReceive::transfer()
+{
+	std::cout << "Receiving..." << std::endl;
+	unsigned long total_received_bytes{0};
+
+	while (!this->is_end && this->shm_ptr->is_init)
+	{
+		lock_mutex();
+		while (this->shm_ptr->data_version_received == this->shm_ptr->data_version)
+		{
+			if (pthread_cond_wait(&(this->shm_ptr->cond), &(this->shm_ptr->mutex)) != 0)
+			{
+				throw std::runtime_error(std::string("ERROR: pthread_cond_wait() send: ") + strerror(errno));
+			}
+		}
+
+		if (this->shm_ptr->shared_mem_size != this->shm_size_in_bytes)
+		{
+			pthread_mutex_destroy(&(this->shm_ptr->mutex));
+			throw std::runtime_error("ERROR: Shared memory size of server and client side are not the same.");
+		}
+
+		if (this->shm_ptr->data_size != 0)
+		{
+			this->file_handler.write_file_shm(this->shm_ptr->data_ap_received, this->shm_ptr->data_size);
+
+			total_received_bytes += this->shm_ptr->data_size;
+			this->shm_ptr->data_version_received++;
+		}
+		else
+		{
+			this->is_end = true;
+			this->shm_ptr->data_version_received++;
+		}
+
+		this->shm_ptr->data_version_received = this->shm_ptr->data_version;
+		receive_cond_broadcast();
+		unlock_mutex();
+		
+		if (pthread_cond_destroy(&(this->shm_ptr->cond)) != 0)
+		{
+			perror("Error at pthread_cond_destroy()\n");
+		}
+
+		if (pthread_mutex_destroy(&(this->shm_ptr->mutex)) != 0)
+		{
+			perror("Error at pthread_mutex_destroy()\n");
```

Comment:
perror can be used only with function that use errno. pthread_mutex_destroy don't use errno but return the error value.

Reply:
And it could be the same with pthread_cond_destroy()

Reply:
Maybe you should not destroy cond and mutex inside the loop?